benches/search_comparison.rs

```diff
@@ -285,3 +255,50 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
 fn search_sorted_vec<'a, T: Ord>(c: &'a Vec<T>, x: T) -> Option<&'a T> {
     c.binary_search(&x).ok().map(|i| &c[i])
 }
+
+/// Generate pseudorandom sequence of numbers
+///
+/// ```rust
+/// assert_eq!(
+///     pseudorandom_iter::<u32>(0, u32::MAX , None).take(16).collect(),
+///     vec![2027808446, 2393657406, 900912232, 833811770, 1061328792, 93432844,
+///     2565420364, 1550801266, 1147887774, 710446582, 3306204668, 500014398,
+///     1140212266, 2163551532, 513205252, 1774545590]
+///
+/// )
+/// ```
+///
+/// ```rust
+/// assert_eq!(
+///     pseudorandom_iter::<u8>(0, u8::MAX, None).take(16).collect(),
+///     vec![250, 200, 36, 20, 178, 78, 4, 152, 174, 8, 128, 198, 166, 206, 156, 80]
+/// );
+/// ```
+///
+/// ```rust
+/// assert_eq!(
+///     pseudorandom_iter::<u8>(0, u8::MAX, Some(32)).take(16).collect(),
+///     vec![27, 9, 5, 20, 18, 15, 5, 24, 15, 8, 1, 6, 7, 15, 29, 17]
+/// );
+fn pseudorandom_iter<T>(
+    mut seed: usize,
+    max: usize,
+    bound: Option<usize>,
+) -> impl Iterator<Item = T>
+where
+    T: TryFrom<usize>,
+    <T as TryFrom<usize>>::Error: core::fmt::Debug,
+{
+    std::iter::from_fn(move || {
+        // LCG constants from https://en.wikipedia.org/wiki/Numerical_Recipes.
+        seed = seed.wrapping_mul(1664525).wrapping_add(1013904223);
+        let r = (seed.wrapping_mul(2)) % max;
```

Comment:
I think this line was actually wrong in the original — it shouldn't have `* 2`. The whole point of only generating even numbers _into_ the collection was so that ~50% of lookups would not have an exact match. But if all the lookups are _also_ even numbers, then that's no longer true.

Reply:
Done